File: system_implementations/golang_solutions/peeking_iterator_284.go
package golang_solutions

type Iterator struct {
}

func (i *Iterator) next() int {
	return 1
}

func (i *Iterator) hasNext() bool {
	return true
}

type IteratorNumber struct {
	number int
}

type PeekingIterator struct {
	iterator   *Iterator
	nextNumber *IteratorNumber
}

func ConstructorPeekingIterator(iter *Iterator) *PeekingIterator {
	p := &PeekingIterator{}
	if iter.hasNext() {
		p.iterator = iter
		p.nextNumber = &IteratorNumber{number: iter.next()}
	}
	return p
}

func (this *PeekingIterator) hasNext() bool {
	return this.nextNumber != nil
}

func (this *PeekingIterator) next() int {
	nextElem := this.nextNumber.number
	if this.iterator.hasNext() {
		this.nextNumber.number = this.iterator.next()
	} else {
		this.nextNumber = nil
	}
	return nextElem
}

func (this *PeekingIterator) peek() int {
	return this.nextNumber.number
}

File: leetcode/system_implementations/java_solutions/PeekingIterator_284.java
package leetcode.system_implementations.java_solutions;

import java.util.Iterator;

public class PeekingIterator_284 implements Iterator<Integer> {

    private final Iterator<Integer> iterator;
    private Integer number;

    public PeekingIterator_284(Iterator<Integer> iterator) {
        // initialize any member here.
        this.iterator = iterator;
        if (this.iterator.hasNext()) this.number = this.iterator.next();
    }

    public Integer peek() {
        return this.number;
    }

    @Override
    public Integer next() {
        Integer nextElem = this.number;
        if (this.iterator.hasNext()) {
            this.number = this.iterator.next();
        } else {
            this.number = null;
        }
        return nextElem;
    }

    @Override
    public boolean hasNext() {
        return this.number != null;
    }
}
